2021q3/copy_staging.adoc

```diff
@@ -0,0 +1,76 @@
+== Changes to UEFI boot on amd64 staging 
+
+Contact: Konstantin Belousov <kib@FreeBSD.org>  
+
+UEFI BIOS on PC provides much more rich and streamlined environment
+for pre-OS programs, but also it imposes some requirements on the
+programs that are executed there, OS loaders in particular.  One of
+such requirement is that loader must coordinate its memory use with
+BIOS, only using memory that was allocated for it.  Even after loader
+handoff to the operating system kernel, there are still some memory
+regions that are reserved by BIOS for different reasons, most knows
+are runtime services code and data.
+
+On the other hand, legacy/CSM BIOS boot works with memory completely
+differently, there are known memory regions that hold BIOS data and
+must be avoided.  Otherwise, the memory is considered free for loader
+and OS to use. [Of course it is not that straightforward, the
+definition of known regions is up to the vendor and there are a lot of
+workarounds there]
+
+BIOS boot puts the kernel and preloaded data (like modules, memory
+disk, CPU microcode update etc) at the contigous physical memory block
+starting at 2M.  This algorithm goes back to how i386 kernel boots.
+
+Also, when preparing to handle the control of to the kernel, loader
+creates quite special temporal mappings, where low 1G of physical
+address space is mapped 1:1 into virtual address space, and then
+repeated for each 1G until the virtual memory end.  Kernel knows about
+its physical location and the temporal mapping, and constructs kernel
+page tables assuming that the physical address of the text is at 2M.
+
+This mechanism of loader to kernel handoff was left unchanged when
+loader got support for UEFI environment.  Loader prepares kernel and
+auxillary preload data in so called staging area while UEFI boot
+services are active, and after EFI_BOOT_SERVICES.ExitBootServices(),
+the temporal mapping is activated and staging area is copied at 2M.
+
+An advantage at that time was that no changes to the kernel were
+needed.  But there are issues, the biggest is that memory at 2M might
+be not free for reuse.  For instance, BIOS runtime code or data might
+be located there.  Or there might be no memory at 2M at all.  Or
+trampoling page table or code, or even some parts of the staging area
+overlapped with the 2M region where staging area is copied.  The
+outcome was a weird and hard to diagnose boot time failures, typically
```

Comment:
"a" doesn't match "failures"
maybe
"hard to diagnose boot time failure, typically a hard hang when the loader started the kernel."